src/services/tokens.ts

```diff
@@ -1,31 +1,39 @@
-import * as JWT from "jsonwebtoken";
+import { decode, Jwt, JwtPayload, verify } from "jsonwebtoken";
 import { JWT_MESSAGE } from "../models/enums";
-import { logEvent } from "../functions/authorizer";
+import { ILogEvent } from "../models/ILogEvent";
+import { checkSignature } from "./signature-check";
 
-export const getValidJwt = (authorizationToken: string): any => {
+interface CVSJWTPayload extends JwtPayload {
+  unique_name: string;
+  preferred_username: string;
+}
+
+export const getValidJwt = async (authorizationToken: string, logEvent: ILogEvent, tenantId: string, clientId: string): Promise<any> => {
   checkFormat(authorizationToken);
 
   authorizationToken = authorizationToken.substring(7); // remove 'Bearer '
-  logEvent.token = authorizationToken;
 
-  const decoded: any = JWT.decode(authorizationToken, { complete: true });
+  const decoded: Jwt | null = decode(authorizationToken, { complete: true });
 
   if (!decoded) {
     throw new Error(JWT_MESSAGE.DECODE_FAILED);
   }
 
+  await checkSignature(authorizationToken, decoded, tenantId, clientId);
+
   let username;
+  const payload = decoded.payload as CVSJWTPayload;
 
-  if (decoded.payload.unique_name) {
-    username = decoded.payload.unique_name;
-  } else if (decoded.payload.preferred_username) {
-    username = decoded.payload.preferred_username;
+  if (decoded.payload) {
```

Comment:
nit: can use variable on line 25

Reply:
Had to alter the structure a little, but that's resolved now